### PR TITLE
Create additional validation checks for use in user creation

### LINKED ISSUE #74: Validations for input createUser
## Description

Implement input validations for the 'createUser' function, encompassing checks for both the UID, username and email.

## Requirements

- [x] Research validations, what kind of validations are best practice working with RESTful API.
- [x] Implemented additional validations to check for empty inputs, inputs of a certain length, and checking of uid, username and emails according to common norms.

### PR DESCRIPTION
**Pull request by oddpharmacy**
# Description

Implement additional checks which cover common scenarios to improve the overall security of the user creation process. This includes length and empty character checks for the uid, username and email user inputs. Error handling for Prisma related errors could be looked into further.

**Note**: The previous PR and branch had some commit issues so I have created a new branch.

# Closes issue(s)

Resolves #74 

# How to test?

### Test using Postman

- Start the server in your local environment. -> `npm run dev`
- Ensure that you have run the necessary prisma scripts, created a local database, and have linked the database url in your env file.
- Open Postman and make a POST req to the _signup_ route. -> `http://localhost:8080/signup`

**_Case 1: A successful request and creation of a new user_**
- Paste the following content as the body to your POST request:
`{
  "username": "totoroo",
  "email": "totoroo@example.com",
  "uid": "aBcDeF1234567890aBcDeF1234--"
}
`
- You should receive an 'id' and 'username' back as a response.

_**Case 2: Invalid inputs entered / Duplicate users**_
- Paste the following content as the body to your POST request:
`{
  "username": "totoroo",
  "email": "totoroo@example.com",
  "uid": "a"
}
`
- You should receive the following error as a response: `Error: Something went wrong. Please check your inputs.`
- Different error messages will pop up depending on the user's inputs. 

# Errors
If you encounter a Typescript error when trying to run the server, make the fix below in _login.service.ts_ on line 14:
`    return loginModel.updateAccessTokenLogin(userId);
`

# Changes include

1. isNotEmptyInput, isValidUID, isValidEmail, and isValidUsername functions.

# Checklist

- [x] I have tested this code

### CODE REVIEW

File: src/controller/signup.controller.ts [hunk 1]
```diff
@@ -8,7 +8,12 @@ export default {
       const id = await serviceSignup.createUser(req.body);
       res.status(201).send(id);
     } catch (error) {
+      let message;
+      if (error instanceof Error) {
+        message = error;
+        res.status(400).send(message.toString());
```
Comment (shonancanuck) [inline src/controller/signup.controller.ts:14]: Should we be sending an explicit error message to the frontend here? Maybe a generic message is better in this case.

File: src/controller/signup.controller.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ export default {
...
```
Comment (shonancanuck) [inline src/controller/signup.controller.ts:14]: I see that most of our errors here are just strings, but we should be careful what we are sending back to the client. Maybe in this case we log the error in the backend and send a generic message to the front?

File: src/service/signup.service.ts [hunk 3]
```diff
@@ -7,21 +7,44 @@ interface SignupData {
   uid: string;
 }
 
+interface CustomError extends Error {
+  code?: string;
+  target?: string;
```
Comment (shonancanuck) [inline src/service/signup.service.ts:12]: Prisma errors sometimes also contain the `errorCode` property (which is different from `code` for some confusing reason). It's probably good to include this too.

**Review by shonancanuck — CHANGES_REQUESTED** [review]
Nice job. I like how you created the CustomError class. It was kind of awkward that the Prisma errors were all separate classes. Just two small points. We can wait to see how the others feel about the response message from the controller if you want.

**Comment by oddpharmacy** [comment]
CLOSED> NEW PR MADE

**Pull request closed without merging**